src/Xrefcheck/Verify.hs

```diff
@@ -269,14 +272,27 @@ verifyReference
                 givenAnchors
           in throwError $ AnchorDoesNotExist anchor similarAnchors
 
-checkExternalResource :: VerifyConfig -> Text -> IO (VerifyResult VerifyError)
-checkExternalResource VerifyConfig{..} link
+checkExternalResource
+  :: VerifyConfig
+  -> MVar (Map Text $ MVar (Either VerifyError ()))
+  -> Text
+  -> IO (VerifyResult VerifyError)
+checkExternalResource VerifyConfig{..} externalCheckedMVar link
   | isIgnored = return mempty
   | doesReferLocalhost = return mempty
   | otherwise = fmap toVerifyRes $ do
-      makeRequest HEAD 0.3 >>= \case
-        Right () -> return $ Right ()
-        Left   _ -> makeRequest GET 0.7
+      externalChecked <- takeMVar externalCheckedMVar
+      case link `M.lookup` externalChecked of
+        Just resultMVar -> putMVar externalCheckedMVar externalChecked >> readMVar resultMVar
+        Nothing -> do
+          resultMVar <- newEmptyMVar
+          let externalChecked' = M.insert link resultMVar externalChecked
+          putMVar externalCheckedMVar externalChecked'
+          result <- makeRequest HEAD 0.3 >>= \case
+            Right () -> return $ Right ()
+            Left   _ -> makeRequest GET 0.7
+          putMVar resultMVar result
```

Comment:
:pogchamp: Yeah I loved that idea!

Reply:
I've understood the logic behind your proposal, but I'm also looking at the pseudocode and can't wrap my head around how to include `verifyReference` itself in the definition of `verifyReferenceRec`... `cachedMap` is used only to store data collected from verifying external links, and the location type of the `Reference` is deduced in `verifyReference`, so that we won't be needing to access the map or insert anything into it if the reference is not an external resource. The fact that the lookup and the insertion operations are performed in `verifyReferenceRec` throws me off a bit. 

Am I correct to assume that, since we're basically redefining the whole traversal logic of the list `[(FilePath, Reference)]` in terms of the recursive function `verifyReferenceRec`, and not via `forConcurrently`, it is also sufficient to redefine **`checkExternalResource`**, where the whole `MVar` business occured, _only_ in terms of the primitive `Map` functions (i.e. no asynchronous actions in the aforementioned function are needed)?

Reply:
Oh right, that's a fair concern.

How about this: decoupling the concurrency/caching logic from the main business logic still seems good unless it gets too complex. 

What if we make our new re-implementation of `forConcurrently` allow optional caching - the map is checked and updated only for external links? 

If we implement this via a generic method, it will have the following signature:

```hs
forConcurrentlyCaching :: Ord cacheKey => [a] -> (a -> NeedsCaching cacheKey) -> (a -> IO b) -> [b]

data NeedsCaching key
  = NoCaching
  | CacheUnderKey key
```

`NoCaching` will be used for local links, and `CacheUnderKey` for external links. Within `forConcurrentlyCaching` we will check and update the map only if the current entry requires caching.

How does that sound?

---

Regarding the second paragraph - yes, this another approach is a replacement for `forConcurrently`. `checkExternalResource` can be reverted back, I think it won't need to know about `Map`, we can just run it only when necessary (`verifyReferenceRec` or `forConcurrentlyCaching` will care about `Map`).

Reply:
Yeah, sounds good! I've already started experimenting with the definition of `forConcurrentlyCaching` some time ago; hopefully, the function's behavior comes out as expected!